### PR TITLE
refactor according html docs and fixed an error

### PR DESCRIPTION
**Pull request by CharlesChen0823**
refactor tokenizer according [docs](https://html.spec.whatwg.org/multipage/parsing.html#script-data-escaped-end-tag-name-state), and fixed an error in character_reference.rs. had verify pass all tokenizer tests.

### CODE REVIEW

**Comment by jaytaph** [comment]
lgtm

**Review by jaytaph — APPROVED** [review]

**Comment by emwalker** [comment]
> and fixed an error in character_reference.rs

The tests pass both on `main` and on this branch.  Can you clarify what error was fixed?

Many thanks for the PR.  I'm looking forward to learning what you're doing here.

**Comment by CharlesChen0823** [comment]
original character_reference.rs in deal with tokenizer/test2.test like `"description":"Hexadecimal entity representing the NUL character",` when input is `FOO&#x0000;ZOO`, the output will be `FOO\uFFFDOO`, but correctly output should be `FOO\uFFFDZOO`.

File: src/html5_parser/tokenizer.rs [hunk 1]
```diff
@@ -352,24 +352,13 @@ impl<'a> Tokenizer<'a> {
...
+                        Element::Utf8(ch) if ch.is_ascii_alphabetic() => {
                             self.current_token = Some(Token::StartTagToken {
                                 name: "".into(),
                                 is_self_closing: false,
                                 attributes: HashMap::new(),
                             });
-
-                            add_to_token_name!(self, ch);
+                            self.stream.unread();
                             self.state = State::TagNameState;
```
[the remark this answers is not in the document]
Comment (CharlesChen0823) [reply src/html5_parser/tokenizer.rs:362]: the photo of your paste is deal with state `script data escapted end tag name state`, the code reference is state `tag open state`

[the remark this answers is not in the document]
Comment (CharlesChen0823) [reply src/html5_parser/tokenizer.rs:362]: and current refactor is according [docs](https://html.spec.whatwg.org/multipage/parsing.html#script-data-escaped-end-tag-name-state), except state `markupdeclarationopenstate` which need finished parser.rs.

[the remark this answers is not in the document]
Comment (emwalker) [reply src/html5_parser/tokenizer.rs:362]: Ah, [I see](https://github.com/gosub-browser/gosub-engine/pull/124/files#diff-71889d349194ad9b083de6854152ca168e9f44f6da435b20a145d3e195407bcdR716-R722).  Thanks for the clarification.

**Review by emwalker — APPROVED** [review]

**Comment by jaytaph** [comment]
@CharlesChen0823 thanks for the pr

**Pull request merged**